splash/benchmark/download_sites.py

```diff
@@ -0,0 +1,107 @@
+#!/usr/bin/env python
+
+"""
+Site downloader script for Splash benchmark suite.
+"""
+
+from argparse import ArgumentParser, ArgumentDefaultsHelpFormatter
+import errno
+import json
+import os
+import re
+import subprocess
+from urlparse import urlsplit
+
+from lxml import html
+
+import w3lib.html
+from splash.benchmark.file_server import serve_files
+from splash.tests.stress import lua_runonce
+
+SCRIPT_HTML = """
+function main(splash)
+splash:set_images_enabled(false)
+splash:go(splash.args.url)
+splash:wait(0.5)
+return {url=splash:url(), html=splash:html()}
+end
+"""
+
+#: This UA is used by httrack to mimic Splash requests when downloading sites.
+USERAGENT = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/534.34 (KHTML, like Gecko) Qt/4.8.1 Safari/534.34"
+
+PORT = 8806
+
+parser = ArgumentParser(description=__doc__,
+                        formatter_class=ArgumentDefaultsHelpFormatter)
+parser.add_argument('--sites-dir', default='sites',
+                    help='Directory for downloaded sites')
+
+
+def preprocess_main_page(sites_dir, url):
+    out = json.loads(lua_runonce(SCRIPT_HTML, url=url,
+                                 splash_args=['--disable-lua-sandbox',
+                                              '--disable-xvfb',
+                                              '--max-timeout=600'],
+                                 timeout=600.,))
+    final_url = urlsplit(out['url'])._replace(query='', fragment='').geturl()
+    if not w3lib.html.get_base_url(out['html']):
```

Comment:
This function does several things, I've updated the docstring to spell them out for clarity.